api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto

```diff
@@ -183,44 +183,219 @@ message JwtHeader {
   string value_prefix = 2;
 }
 
-// This is the Envoy HTTP filter config for JWT authentication.
-// [#not-implemented-hide:]
-message JwtAuthentication {
-  // List of JWT rules to valide.
-  repeated JwtRule rules = 1;
+// Specify a required provider with audiences.
+message ProviderWithAudiences {
+  // Specify a required provider name.
+  string provider_name = 1;
 
-  // If true, the request is allowed if JWT is missing or JWT verification fails.
-  // Default is false, a request without JWT or failed JWT verification is not allowed.
-  bool allow_missing_or_failed = 2;
+  // This field overrides the one specified in the JwtProvider.
+  repeated string audiences = 2;
+}
 
-  // This field lists the patterns allowed to bypass JWT verification. This only applies when
-  // `allow_missing_or_failed_jwt` is false. Under this config, if a request doesn't have JWT, it
-  // will be rejected. But some requests still needed to be forwarded without JWT, such as OPTIONS
-  // for CORS and some health checking paths.
+// This message specifies a Jwt requirement. An empty message means JWT verification is not
+// required. Here are some config examples:
+//
+// .. code-block:: yaml
+//
+//  # Example 1: not required with an empty message
+//
+//  # Example 2: require A
+//  provider_name: "provider-A"
+//
+//  # Example 3: require A or B
+//  requires_any:
+//    requirements:
+//      - provider_name: "provider-A"
+//      - provider_name: "provider-B"
+//
+//  # Example 4: require A and B
+//  requires_all:
+//    requirements:
+//      - provider_name: "provider-A"
+//      - provider_name: "provider-B"
+//
+//  # Example 5: require A and (B or C)
+//  requires_all:
+//    requirements:
+//      - provider_name: "provider-A"
+//      - requires_any:
+//        requirements:
+//          - provider_name: "provider-B"
+//          - provider_name: "provider-C"
+//
+//  # Example 6: require A or (B and C)
+//  requires_any:
+//    requirements:
+//      - provider_name: "provider-A"
+//      - requires_all:
+//        requirements:
+//          - provider_name: "provider-B"
+//          - provider_name: "provider-C"
+//
+message JwtRequirement {
+  oneof requires_type {
+    // Specify a required provider name.
+    string provider_name = 1;
+
+    // Specify a required provider with audiences.
+    ProviderWithAudiences provider_and_audiences = 2;
+
+    // Specify list of JwtRequirement. Their results are OR-ed.
+    // If any one of them passes, the result is passed.
+    JwtRequirementOrList requires_any = 3;
+
+    // Specify list of JwtRequirement. Their results are AND-ed.
+    // All of them must pass, if one of them fails or missing, it fails.
+    JwtRequirementAndList requires_all = 4;
+  }
+}
+
+// This message specifies a list of RequiredProvider.
+// Their results are OR-ed; if any one of them passes, the result is passed
+message JwtRequirementOrList {
```

Comment:
Out of curiosity, would this work if we just inline it into the above fields? E.g. if we had `repeasted JwtRequirement requires_any = 3 [constraints];`? I'm not familiar with the proto rules on self-reference.

Reply:
Not, it will not work,  proto "oneof" doesn't "repeated" field.